ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -622,6 +622,12 @@ static const upb_pbdecodermethod *msgdef_decodermethod(Descriptor* desc) {
   return desc->fill_method;
 }
 
+// Callback invoked by upb if any error occurs during parsing or serialization.
+static bool env_error_func(void* ud, const upb_status* status) {
+  const char* message = (const char*)ud;
+  rb_raise(rb_eRuntimeError, message, upb_status_errmsg(status));
```

Comment:
rb_raise() does longjmp IIRC? If so, it will leak the environment. Seems like you either want to call upb_env_uninit() here or anchor the upb_env to a Ruby object that will get a finalizer call where you can call upb_env_uninit().